oop.md

```diff
@@ -0,0 +1,179 @@
+  private static instance: AxiosInstance;
+
+  private constructor() { }
+
+  public static getInstance(): AxiosInstance {
+    if (!this.instance) this.instance = axios.create({
+      baseURL: `${process.env.NEXT_PUBLIC_API_HOST}/${process.env.NEXT_PUBLIC_API_VERSION}`,
+      headers: {
+        'Content-Type': 'application/json',
+        Authorization: `Bearer ${process.env.NEXT_PUBLIC_API_TOKEN}`,
+      },
+    });
+
+    return this.instance;
+  }
+}
+
+export default SingletonInstance.getInstance();
+
+```
+
+# Builder
+
+I have also used redux here, so I have here an example of the using built-in realization of the builder GoF pattern which allows constructing complex objects step by step. In the redux we have ability to work with extra reducers by using builder. We can add as much cases as we want with using addcase() method.
+
+
+#####  lessons slice
+
+```js
+import { createSlice } from '@reduxjs/toolkit';
+
+import { Status } from '../types';
+
+import { fetchLessons } from './asyncActions';
+import { LessonsSliceState } from './types';
+
+const initialState: LessonsSliceState = {
+  status: Status.LOADING,
+  lessons: [],
+  courseInfo: null,
+};
+
+const lessonsSlice = createSlice({
+  name: 'lessons',
+  initialState,
+  reducers: {},
+  extraReducers: builder => {
+    builder.addCase(fetchLessons.pending, state => {
+      delete state.Error;
+      state.status = Status.LOADING;
+    });
+    builder.addCase(fetchLessons.fulfilled, (state, action) => {
+      const { lessons, ...courseInfo } = action.payload;
+      state.lessons = lessons;
+      state.courseInfo = courseInfo;
+      state.status = Status.SUCCESS;
+    });
+    builder.addCase(fetchLessons.rejected, (state, action) => {
+      state.status = Status.ERROR;
+      state.Error = action.error.message;
+    });
+  },
+});
+
+export default lessonsSlice.reducer;
+```
+
+#####  courses slice
+
+```js
+import { createSlice } from '@reduxjs/toolkit';
+
+import { Status } from '../types';
+
+import { fetchCourses } from './asyncActions';
+import { CoursesSliceState } from './types';
+
+const initialState: CoursesSliceState = {
+  status: Status.LOADING,
+  courses: [],
+  currentPage: 1,
+};
+
+const coursesSlice = createSlice({
+  name: 'courses',
+  initialState,
+  reducers: {
+    setCurrentPage(state, action) {
+      state.currentPage = action.payload;
+    },
+  },
+  extraReducers: builder => {
+    builder.addCase(fetchCourses.pending, state => {
+      delete state.Error;
+      state.status = Status.LOADING;
+    });
+    builder.addCase(fetchCourses.fulfilled, (state, action) => {
+      state.courses = action.payload;
+      state.status = Status.SUCCESS;
+    });
+    builder.addCase(fetchCourses.rejected, (state, action) => {
+      state.status = Status.ERROR;
+      state.Error = action.error.message;
+    });
+  },
+});
+
+export const { setCurrentPage } = coursesSlice.actions;
+
+export default coursesSlice.reducer;
+```
+# Observer
```

Comment:
Yes, I agree, in Redux, the store can be seen as the observable object, and the components that use useSelector and useDispatch can be seen as the observers. The store notifies the components of state changes, and the components notify the store of user actions.